keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,987 @@
+
+#### K8s Update Tightens CRD validation
+
+In this case the user has a CRD with a field with the following schema:
+
+```yaml
+type: string
+format: dns1123subdomain
+```
+
+Kubernetes does not recognize this as a format. Today, Kubernetes treats
+[unrecognized formats as a no-op](https://github.com/kubernetes/kubernetes/blob/f9f9e7177a328ae47bac6b823d57a99dfbfd310b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/formats.go#L53-L65) and ignores them. So users would be able
+to write fields not conforming to this format to custom resources.
+
+If Kubernetes wishes to support this format, users workflows will break upon
+upgrade until the offending resources and identified and repaired.
+
+With this CRD Validation Ratcheting feature, Kubernetes would be able to
+adopt new format strings safely without breaking user workflows.
+
+#### K8s Update Widens CRD Validation
+
+In this case the user has a CRD with a field with following schema:
+
+```yaml
+type: string
+format: byte
+```
+
+In k8s 1.27 and below, this format disallows empty strings due to an implementation detail. However, in native types like `Secret`, byte is disallowed. Is Kubernetes had an update to CRDs to allow empty string for `byte`, this widens the validation for those CRDS which use it.
+
+This is fine when upgrading to new versions if Kubernetes: all existing resources still pass validation.
+
+The user may still experience issues when downgrading, for example if they had unreletated issues with the new Kubernetes update:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers fail to update some resources, since they fail validation.
+
+If the CRD Validation Ratcheting feature were enabled the following occurs:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers are allowed to update their resources, as long as all their changes are valid.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+This KEP is scoped to CRDs, but if [KEP-XXX: Declarative Validation]() is implemented then we should expect it to also affect native type validation.
+
+### Risks and Mitigations
+
+Help Wanted
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
```

Comment:
There are two correlation issues that are worth spelling out:

- What happens if a valid string is pretended to an atomic list containing invalid strings
- A field path containing a listType=map will still use a positional index, but the old and new value should be correlated using the map key, no the list position. How is this handled?

Reply:
For the atomic list case I think we will be unable to ratchet the unmodified items.

For the listType=map case we can lookup the map key in the new object, and see if the old object has it.

I've added these details to the KEP

Reply:
Another edge-case: take `list-type=set` with changed order and strengthened validation, e.g.
```yaml
type: array
x-k-list-type: set
items:
  type: string
  pattern: `^x` # this is new
```
Updating `a,b,c` to `a,c,b`. This is not supposed to fail I would assume.

And you can make this even more complex in a number of other dimensions:

E.g. by nesting those properties in a way that the error path is definitely not item path:

```yaml
type: array
x-k-list-type: set
items:
  type: string
or: # this is new
- false: true
- items:
    pattern: `^x` 
```

This suggests that we need a normal form of some kind.

Reply:
Added details of ValidateListsSetsAndMaps and how normal form will be used